core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java

```diff
@@ -0,0 +1,105 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model.identity;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+
+/**
+ * A source of instance identity.
+ *
+ * @param <PUB>  the type of public key.
+ * @param <PRIV> the type of private key.
+ * @since FIXME
+ */
+public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV extends PrivateKey> implements
+        ExtensionPoint {
+    /**
+     * Gets the {@link KeyPair} that comprises the instance identity.
+     *
+     * @return the {@link KeyPair} that comprises the instance identity. {@code null} could technically be returned in
+     * the event that a keypair could not be generated, for example if the specific key type of this provider
+     * is not permitted at the required length by the JCA policy.
+     */
+    @CheckForNull
+    public abstract KeyPair getKeyPair();
+
+    /**
+     * Shortcut to {@link KeyPair#getPublic()}.
+     *
+     * @return the public key. {@code null} if {@link #getKeyPair()} is {@code null}.
+     */
+    @CheckForNull
+    public PUB getPublicKey() {
+        KeyPair keyPair = getKeyPair();
+        return keyPair == null ? null : (PUB) keyPair.getPublic();
+    }
+
+    /**
+     * Shortcut to {@link KeyPair#getPrivate()}.
+     *
+     * @return the private key. {@code null} if {@link #getKeyPair()} is {@code null}.
+     */
+    @CheckForNull
+    public PRIV getPrivateKey() {
+        KeyPair keyPair = getKeyPair();
+        return keyPair == null ? null : (PRIV) keyPair.getPrivate();
+    }
+
+    /**
+     * Gets the self-signed {@link X509Certificate} that is associated with this identity. The certificate
+     * will must be currently valid. Repeated calls to this method may result in new certificates being generated.
+     *
+     * @return the certificate. {@code null} if {@link #getKeyPair()} is {@code null}.
```

Comment:
Yes. If you have a key pair you should be able to generate a self signed certificate from the key pair